File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
import sys
from itertools import combinations

input = sys.stdin.readline
write = sys.stdout.write

def get_diff(start_team, n, nlist):
    link_team = [i for i in range(n) if i not in start_team] 
    
    start_score = sum(nlist[p1][p2] + nlist[p2][p1] for p1, p2 in combinations(start_team, 2))
    link_score = sum(nlist[p3][p4] + nlist[p4][p3] for p3, p4 in combinations(link_team, 2))
    
    return abs(start_score - link_score)

def dfs(idx, count, n, nlist, start_team, min_diff): 
    if count == n // 2:
        diff = get_diff(start_team, n, nlist)
        min_diff[0] = min(min_diff[0], diff)
        return

    for i in range(idx, n):
        start_team.append(i)
        dfs(i + 1, count + 1, n, nlist, start_team, min_diff) 
        start_team.pop()

def main():
    n = int(input())
    if not n: return
    if n % 2 != 0: return
    
    nlist = [list(map(int, input().split())) for _ in range(n)]
    
    min_diff = [1000000000]
    
    dfs(1, 1, n, nlist, [0], min_diff)
    
    write(str(min_diff[0]) + "\n")

if __name__ == "__main__":
    main()

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.cc
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int n;
int s[20][20];
int visited[20];
int min_diff = 1000000000;

void calculate() {
	int start_score = 0;
	int link_score = 0;

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (visited[i] && visited[j]) {
				start_score += s[i][j];
			}
			else if (!visited[i] && !visited[j]) {
				link_score += s[i][j];
			}
		}
	}

	int diff = abs(start_score - link_score);
	if (diff < min_diff) min_diff = diff;
}

void dfs(int idx, int count) {
	if (count == n / 2) {
		calculate();
		return;
	}

	for (int i = idx; i < n; i++) {
		if (!visited[i]) {
			visited[i] = 1;

			dfs(i + 1, count + 1);

			visited[i] = 0;
		}
	}
}

int main(void) {
	if (scanf("%d", &n) != 1) return 0;

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (scanf("%d", &s[i][j]) != 1) break;
		}
	}

	visited[0] = 1;
	dfs(1, 1);

	printf("%d\n", min_diff);

	return 0;
}